arduino-ide-extension/src/browser/contributions/update-indexes.ts

```diff
@@ -0,0 +1,193 @@
+            typesToUpdate
+          )} with date time: ${updatedAt}.`
+        );
+        return this.coreService.updateIndex({ types: typesToUpdate });
+      }
+    });
+  }
+
+  private async needsIndexUpdate(
+    type: IndexType,
+    now: string,
+    force = false
+  ): Promise<IndexType | false> {
+    if (force) {
+      console.debug(
+        `[update-indexes]: Update for index type: '${type}' was forcefully requested.`
+      );
+      return type;
+    }
+    const lastUpdateIsoDateTime = await this.getLastUpdateDateTime(type);
+    if (!lastUpdateIsoDateTime) {
+      console.debug(
+        `[update-indexes]: No last update date time was persisted for index type: '${type}'. Index update is required.`
+      );
+      return type;
+    }
+    const lastUpdateDateTime = Date.parse(lastUpdateIsoDateTime);
+    if (Number.isNaN(lastUpdateDateTime)) {
+      console.debug(
+        `[update-indexes]: Invalid last update date time was persisted for index type: '${type}'. Last update date time was: ${lastUpdateDateTime}. Index update is required.`
+      );
+      return type;
+    }
+    const diff = new Date(now).getTime() - lastUpdateDateTime;
+    const needsIndexUpdate = diff >= this.threshold;
+    console.debug(
+      `[update-indexes]: Update for index type '${type}' is ${
+        needsIndexUpdate ? '' : 'not '
+      }required. Now: ${now}, Last index update date time: ${new Date(
+        lastUpdateDateTime
+      ).toISOString()}, diff: ${diff} ms, threshold: ${this.threshold} ms.`
+    );
+    return needsIndexUpdate ? type : false;
+  }
+
+  private async getLastUpdateDateTime(
+    type: IndexType
+  ): Promise<string | undefined> {
+    const key = this.storageKeyOf(type);
+    return this.localStorage.getData<string>(key);
+  }
+
+  private async setLastUpdateDateTime(
+    type: IndexType,
+    updatedAt: string
+  ): Promise<void> {
+    const key = this.storageKeyOf(type);
+    return this.localStorage.setData<string>(key, updatedAt).finally(() => {
+      console.debug(
+        `[update-indexes]: Updated the last index update date time of '${type}' to ${updatedAt}.`
+      );
+    });
+  }
+
+  private storageKeyOf(type: IndexType): string {
+    return `index-last-update-time--${type}`;
+  }
+
+  private get threshold(): number {
+    return 4 * 60 * 60 * 1_000; // four hours in millis
+  }
+}
+export namespace UpdateIndexes {
+  export namespace Commands {
+    export const UPDATE_INDEXES: Command & { label: string } = {
+      id: 'arduino-update-indexes',
+      label: nls.localize(
+        'arduino/updateIndexes/updateIndexes',
+        'Update Indexes'
+      ),
+      category: 'Arduino',
+    };
+    export const UPDATE_PLATFORM_INDEX: Command & { label: string } = {
+      id: 'arduino-update-platform-index',
+      label: nls.localize(
+        'arduino/updateIndexes/updatePlatformIndex',
+        'Update Platform Index'
```

Comment:
```suggestion
        'Update Package Index'
```

"Package index" is the correct terminology here:

https://arduino.github.io/arduino-cli/dev/package_index_json-specification/

A package is a container for platforms and tools, so "platform index" is not a very accurate description of this thing. Since tools are usually only an asset for the platforms in the index, users are typically focused on platforms when thinking of a package index. However, any platform can reference a tool dependency from any other package, so a package index may serve solely for the distribution of tools ([example](https://forum.arduino.cc/t/error-during-installation-index-error-could-not-find-referenced-tool/1036410)).

In the primary user interface, package indexes may also be referred to by the more user understandable "Boards Manager URL" term. In this case of the advanced interface, and especially since there is no established "Library Manager URL" term to use for the parallel library index command, I think "package index" is the right choice.

Reply:
<img width="653" alt="Screen Shot 2022-10-04 at 10 17 56" src="https://user-images.githubusercontent.com/1405703/193769895-97db4bb9-0f9d-49a8-8658-95c099a6948b.png">